### PR TITLE
Add type safe rendering methods

### PR DESCRIPTION
**Pull request by ddunn2**
## Description
Adds type safe templating methods to Kitura. Two new render methods were added to `RouterResponse.swift`, one for an Encodable type and another for a tuple of `(Identifier, Encodable)`. Both of the aforementioned methods call the internal render method added to `Router.swift`. 

A suite of tests has been added as well.  

## Motivation and Context
See this [proposal](https://github.com/ddunn2/evolution/blob/typesafe-template/Kitura/Proposals/0006-type-safe-templating.md)

## How Has This Been Tested?
New tests have been added covering the new API. These tests (along with the entire suite) have been run on MacOS and Linux.

### CODE REVIEW

File: Sources/Kitura/TemplatingError.swift [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import Foundation
 
 enum TemplatingError: Swift.Error {
```
Comment (ianpartridge) [inline Sources/Kitura/TemplatingError.swift]: I'm confused, isn't the default access level `internal`? How are we managing to throw this from a public method at the moment??

File: Sources/Kitura/TemplatingError.swift [hunk 2]
```diff
@@ -17,6 +17,7 @@
...
+    case noKeyProvidedForArrayType
```
Comment (ianpartridge) [inline Sources/Kitura/TemplatingError.swift]: Isn't this SemVer major?

File: Sources/Kitura/RouterResponse.swift [hunk 3]
```diff
@@ -363,6 +364,58 @@ public class RouterResponse {
...
+    /// - Throws: TemplatingError if no file extension was specified or there is no template engine defined for the extension.
+    /// - Returns: this RouterResponse.
+    @discardableResult
+    public func render<T: Encodable>(_ resource: String, with value: T, forKey key: String? = nil,
+                       options: RenderingOptions = NullRenderingOptions()) throws -> RouterResponse {
+        
+        //Throw an error is an array is passed without providing a key.
+        if key == nil {
+            let mirror = Mirror(reflecting: value)
+            if mirror.displayStyle == .collection {
```
[the remark this answers is not in the document]
Comment (ddunn2) [reply Sources/Kitura/RouterResponse.swift]: We do not, as the dictionary provides it's own key for the template engines to use.

File: Sources/Kitura/Router.swift [hunk 4]
```diff
@@ -203,30 +203,61 @@ public class Router {
...
+                         options: RenderingOptions = NullRenderingOptions()) throws -> String {
+        let (optionalFileExtension, resourceWithExtension) = calculateExtension(template: template)
+        
+        let templateEngine = try validateTemplate(template: template, optionalExtension: optionalFileExtension)
+        let absoluteFilePath = buildAbsoluteFilePath(with: resourceWithExtension)
+        
+        return try templateEngine.render(filePath: absoluteFilePath, with: value, forKey: key, options: options, templateName: resourceWithExtension)
+    }
+    
+    private func validateTemplate(template: String, optionalExtension: String?) throws -> TemplateEngine {
```
Comment (ianpartridge) [inline Sources/Kitura/Router.swift]: As this returns a TemplateEngine a better name might be something like `getTemplateEngineForTemplate()`?

File: Tests/KituraTests/TestTemplateEngine.swift [hunk 5]
```diff
@@ -236,12 +263,167 @@ class TestTemplateEngine: KituraTest {
...
+        } catch {
+            XCTFail("Error during render \(error)")
+        }
+    }
+    
+    func testCodableRender() {
+        let router = Router()
+        let templateEngine = MockTemplateEngine()
+        router.add(templateEngine: templateEngine)
+//        router.setDefault(templateEngine: templateEngine)
```
Comment (ianpartridge) [inline Tests/KituraTests/TestTemplateEngine.swift]: Remove this?

File: Tests/KituraTests/TestTemplateEngine.swift [hunk 6]
```diff
@@ -136,6 +145,24 @@ class TestTemplateEngine: KituraTest {
             }
         }
     }
+    
+    private func setupRouterForCodableRendering(_ router: Router, options: RenderingOptions? = nil) {
+        router.setDefault(templateEngine: MockTemplateEngine())
+        
+        router.get("/render") { _, response, next in
+            do {
+                if let options = options {
```
Comment (ianpartridge) [inline Tests/KituraTests/TestTemplateEngine.swift]: This `if` is unneeded - we can just do:

```swift
try response.render("test.mock", with: self.person, forKey: "", options: options)
next()
```

File: Sources/Kitura/RouterResponse.swift [hunk 7]
```diff
@@ -363,6 +364,56 @@ public class RouterResponse {
...
+    ///
+    /// - Parameter resource: the resource name without extension.
+    /// - Parameter with: an array of tuples of type (Identifier, Encodable). The Encodable values are used to generate the content.
+    /// - Parameter forKey: A value used to match the Encodable values to the correct variable in a template file.
+    ///                                 The `forKey` value should match the desired variable in the template file.
+    /// - Parameter options: rendering options, specific per template engine
+    /// - Throws: TemplatingError if no file extension was specified or there is no template engine defined for the extension.
+    /// - Returns: this RouterResponse.
+    @discardableResult
+    public func render<I: Identifier, T: Encodable>(_ resource: String, with values: [(I, T)], forKey key: String,
```
Comment (ianpartridge) [inline Sources/Kitura/RouterResponse.swift]: This needs some kind of test please.

**Comment by ddunn2** [comment]
Review comments have been addressed

File: Sources/Kitura/RouterResponse.swift [hunk 8]
```diff
@@ -363,6 +364,56 @@ public class RouterResponse {
...
+    /// - Throws: TemplatingError if no file extension was specified or there is no template engine defined for the extension.
+    /// - Returns: this RouterResponse.
+    @discardableResult
+    public func render<T: Encodable>(_ resource: String, with value: T, forKey key: String? = nil,
+                       options: RenderingOptions = NullRenderingOptions()) throws -> RouterResponse {
+        
+        //Throw an error if an array is passed without providing a key.
+        if key == nil {
+            let mirror = Mirror(reflecting: value)
+            if mirror.displayStyle == .collection {
```
Comment (ianpartridge) [inline Sources/Kitura/RouterResponse.swift]: Need to consider `.set` too.

File: Sources/Kitura/RouterResponse.swift [hunk 9]
```diff
@@ -363,6 +364,56 @@ public class RouterResponse {
...
+    ///                                 The `forKey` value should match the desired variable in the template file.
+    /// - Parameter options: rendering options, specific per template engine
+    /// - Throws: TemplatingError if no file extension was specified or there is no template engine defined for the extension.
+    /// - Returns: this RouterResponse.
+    @discardableResult
+    public func render<T: Encodable>(_ resource: String, with value: T, forKey key: String? = nil,
+                       options: RenderingOptions = NullRenderingOptions()) throws -> RouterResponse {
+        
+        //Throw an error if an array is passed without providing a key.
+        if key == nil {
```
Comment (ianpartridge) [inline Sources/Kitura/RouterResponse.swift]: I think this needs to be pushed down into Stencil and Mustache, as discussed.

**Pull request merged**